k-distribution/k-tutorial/1_basic/16_collections/README.md

```diff
@@ -1 +1,344 @@
+`SET` module. Like maps and lists, this type is not (currently) polymorphic.
+Like `Map` and `List`, all `Sert` terms are sets of terms of the `KItem` sort.
+
+A `Set` pattern has the exact same restrictions as a `Map` pattern, except that
+its elements are treated like keys, and there are no values. It has the same
+performance characteristics as well. However, syntactically it is more similar
+to the `List` sort: An empty `Set` is represented by `.Set`, but a set element
+is represented by the `SetItem` symbol.
+
+Matching behaves similarly to the `Map` sort: a set pattern with no variables
+of sort `Set` will match if the set has exactly as many bindings as `SetItem`
+symbols, and if each element pattern matches one distinct element in the set.
+A set with a variable of sort `Set` also matches any superset of such a set.
+As with map, the elements left over will be bound to the `Set` variable (or
+`.Set` if no elements are left over).
+
+Like `Map`, the `...` syntax on a set is syntactic sugar for an anonymous
+variable of sort `Set`.
+
+Here is an example of a simple modification to `LESSON-16-A` which uses a `Set`
+to ensure that variables are never declared more than once. In practice, you
+would likely just use the `in_keys` symbol over maps to test for this, but
+it's still useful as an example of sets in practice:
+
+```k
+module LESSON-16-C-SYNTAX
+  imports LESSON-16-A-SYNTAX
+endmodule
+
+module LESSON-16-C
+  imports LESSON-16-C-SYNTAX
+  imports BOOL
+
+  configuration <T>
+                  <k> $PGM:Pgm </k>
+                  <state> .Map </state>
+                  <declared> .Set </declared>
+                </T>
+
+  // declaration sequence
+  rule <k> D:Decl P:Pgm => D ~> P ...</k>
+  rule <k> .Pgm => . ...</k>
+
+  // variable declaration
+  rule <k> int X:Id = I:Int ; => . ...</k>
+       <state> STATE => STATE [ X <- I ] </state>
+       <declared> D => D SetItem(X) </declared>
+    requires notBool X in D
+
+  // variable lookup
+  rule <k> X:Id => I ...</k>
+       <state>... X |-> I ...</state>
+       <declared>... SetItem(X) ...</declared>
+
+  syntax Bool ::= isKResult(K) [symbol, function]
+  rule isKResult(_:Int) => true
+  rule isKResult(_) => false [owise]
+endmodule
+```
+
+Now if we `krun` a program containing duplicate declarations, it will get
+stuck on the declaration.
+
+## Exercises
+
+1. Modify your solution to Lesson 1.14, Problem 2 and introduce the sorts
+`Decls`, `Decl`, and `Stmt` which include variable and function declaration
+(without function parameters), and return and assignment statements, as well
+as call expressions. Use `List` and `Map` to implement these operators, making
+sure to consider the interactions between components, such as saving and
+restoring the environment of variables at each call site. Don't worry about
+local function definitions or global variables for now. Make sure to test the
+resulting interpreter.
```

Comment:
This seems like a bit of a stretch in difficulty. Are you intending that they store previous call-states? Like are you intending something like:
```
<k> call f(...) ~> REST => FBODY </k>
<callState> .List => ListItem(REST | MEM | DECL) ... </callState>
<state> MEM => .Map </state>
<declared> DECL => .Set </declared>
```

Have we introduced the concept of storing the continuation (the `~> REST` portion)?

I think we'll need to provide more guidance about how to complete this. I think you'll have to at least mention something like "Use a new cell called `<callStack>` which is a list of call-states, where a call-state consists of the current continuation, the current state, and the current declared variable." or the like.

Reply:
I'm not sure I agree with the claim that this will be too hard. It's essentially just the combination of the previous examples in this lesson. The only really new component involves saving and restoring certain cells when a new call frame is created, something that we already explained how to do earlier in this lesson.